GSB: use a request for the structural type of type aliases

xymus opened the pull request.

xymus committed:
GSB: use StructuralTypeRequest for type aliases

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test

slavapestov commented on lib/AST/TypeCheckRequests.cpp:
```diff
@@ -546,3 +546,27 @@ bool DefaultTypeRequest::getPerformLocalLookup(const KnownProtocolKind knownProt
+
+Optional<Type> StructuralTypeRequest::getCachedResult() const {
+  auto decl = std::get<0>(getStorage());
+  if (decl->getUnderlyingTypeLoc().getType())
```
We don't want to cache it here because this is where the interface type will be stored. Instead, you can change CacheKind::SeparatelyCached to CacheKind::Cached so that the request-evaluator uses its own caching.

xymus commented:
@swift-ci Please smoke test

slavapestov reviewed (approved):
Looks good!

xymus committed:
Sema: use the underlying type when available instead of the structural type request

The underlying type may have been set on a validated type alias or when deserialized.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
Fix looks good

slavapestov commented:
@swift-ci Please smoke test macOS

xymus committed:
Sema: intro a request evaluator to get the structural type of type aliases

This request evaluator doesn't rely on the previous execution of
a compilation phase but it caches the result in a compatible way.

xymus committed:
Sema: make StructuralTypeRequest use CacheKind::Cached

xymus commented:
@swift-ci Please smoke test macOS

xymus commented:
@swift-ci Please smoke test Linux

xymus commented:
@swift-ci Please test source compatibility

slavapestov commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -1008,3 +1008,11 @@ RequirementRequest::evaluate(Evaluator &evaluator,
+                                       TypeAliasDecl *D) const {
+  auto typeRepr = D->getUnderlyingTypeLoc().getTypeRepr();
+  auto resolution = TypeResolution::forStructural(D);
+  return resolution.resolveType(typeRepr, None);
```
You need to use the same options as in `validateTypeAliasType()`:

```
  TypeResolutionOptions options(
    (typeAlias->getGenericParams() ?
     TypeResolverContext::GenericTypeAliasDecl :
     TypeResolverContext::TypeAliasDecl));

  if (!typeAlias->getDeclContext()->isCascadingContextForLookup(
        /*functionsAreNonCascading*/true)) {
     options |= TypeResolutionFlags::KnownNonCascadingDependency;
  }
```

For now you can just copy and paste that code, we'll factor it out later.

slavapestov replied:
... however a generic type alias will never end up here, so you remove that part. In fact, can you `assert(!getGenericParams())` in `TypeAliasDecl::getStructuralType()`?

xymus committed:
Sema: restrict the options to resolve the structural type of type aliases

xymus commented:
@swift-ci Please smoke test macOS

xymus commented:
@swift-ci Please smoke test Linux

xymus commented:
@swift-ci Please test source compatibility

slavapestov commented:
It looks like expandConformanceRequirement() needs to skip generic type aliases altogether.

slavapestov commented:
... and generic types in general. They're not something the GSB tries to model anyway.

xymus committed:
GSB: ignore generic types in expandConformanceRequirement

Fixes crashes in 28437-swift-typechecker-validatedecl.swift (from previous commit)
and the compiler crasher 28861-gpdecl-getdepth-generictypeparamdecl-
invaliddepth-parameter-hasnt-been-validated.swift.

xymus commented:
@swift-ci Please smoke test macOS

xymus commented:
@swift-ci Please smoke test Linux

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please smoke test Linux
